barretenberg/cpp/src/barretenberg/proof_system/circuit_builder/generated/AvmMini_trace.cpp

```diff
@@ -0,0 +1,346 @@
+
+    auto insertionIndex = std::lower_bound(memTrace.begin(), memTrace.end(), newMemEntry, compareMemEntries);
+    memTrace.insert(insertionIndex, newMemEntry);
+}
+
+// Memory operations need to be performed before the addition of the corresponding row in
+// mainTrace, otherwise the m_clk value will be wrong. This applies to:
+//       loadAInMemTrace, loadBInMemTrace, loadCInMemTrace
+//       storeAInMemTrace, storeBInMemTrace, storeCInMemTrace
+
+void AvmMiniTraceBuilder::loadAInMemTrace(uint32_t addr, FF val)
+{
+    insertInMemTrace(static_cast<uint32_t>(mainTrace.size()), 0, addr, val, false);
+}
+
+void AvmMiniTraceBuilder::loadBInMemTrace(uint32_t addr, FF val)
+{
+    insertInMemTrace(static_cast<uint32_t>(mainTrace.size()), 1, addr, val, false);
+}
+
+void AvmMiniTraceBuilder::loadCInMemTrace(uint32_t addr, FF val)
+{
+    insertInMemTrace(static_cast<uint32_t>(mainTrace.size()), 2, addr, val, false);
+}
+
+void AvmMiniTraceBuilder::storeAInMemTrace(uint32_t addr, FF val)
+{
+    insertInMemTrace(static_cast<uint32_t>(mainTrace.size()), 3, addr, val, true);
+}
+
+void AvmMiniTraceBuilder::storeBInMemTrace(uint32_t addr, FF val)
+{
+    insertInMemTrace(static_cast<uint32_t>(mainTrace.size()), 4, addr, val, true);
+}
+
+void AvmMiniTraceBuilder::storeCInMemTrace(uint32_t addr, FF val)
+{
+    insertInMemTrace(static_cast<uint32_t>(mainTrace.size()), 5, addr, val, true);
+}
+
+// Addition over finite field with direct memory access.
+void AvmMiniTraceBuilder::add(uint32_t s0, uint32_t s1, uint32_t d0)
+{
+    // a + b = c
+    FF a = ffMemory.at(s0);
+    FF b = ffMemory.at(s1);
+    FF c = a + b;
+    ffMemory.at(d0) = c;
+
+    auto clk = mainTrace.size();
+
+    // Loading into Ia
+    loadAInMemTrace(s0, a);
+
+    // Loading into Ib
+    loadBInMemTrace(s1, b);
+
+    // Storing from Ic
+    storeCInMemTrace(d0, c);
+
+    mainTrace.push_back(Row{
+        .avmMini_clk = clk,
+        .avmMini_subop = FF(1),
+        .avmMini_ia = a,
+        .avmMini_ib = b,
+        .avmMini_ic = c,
+        .avmMini_mem_op_a = FF(1),
+        .avmMini_mem_op_b = FF(1),
+        .avmMini_mem_op_c = FF(1),
+        .avmMini_rwc = FF(1),
+        .avmMini_mem_idx_a = FF(s0),
+        .avmMini_mem_idx_b = FF(s1),
+        .avmMini_mem_idx_c = FF(d0),
+    });
+};
+
+// CALLDATACOPY opcode with direct memory access, i.e.,
+// M_F[d0:d0+s1] = M_calldata[s0:s0+s1]
+// Simplified version with exclusively memory store operations and
+// values from M_calldata passed by an array and loaded into
+// intermediate registers.
+// Assume that caller passes callDataMem which is large enough so that no out-of-bound
+// memory issues occur.
+// TODO: Implement the indirect memory version (maybe not required)
+// TODO: taking care of intermediate register values consistency and propagating their
+// values to the next row when not overwritten.
+void AvmMiniTraceBuilder::callDataCopy(uint32_t s0, uint32_t s1, uint32_t d0, std::vector<FF> const& callDataMem)
```

Comment:
So the idea is, we simply fill the intermediate registers with values right now and then store those values in memory. There is no in-machine calldata memory right now, right?

Reply:
Correct.